pytest_ethereum/testing.py

```diff
@@ -0,0 +1,96 @@
+from typing import Any, Dict
+
+from eth_tester.exceptions import TransactionFailed
+import pytest
+from web3.contract import ContractEvent
+
+from pytest_ethereum._utils.abi import merge_args_and_kwargs
+from pytest_ethereum.exceptions import LogError
+
+TxReceipt = Dict[str, Any]
+
+
+class Log:
+    def __init__(
+        self, contract_event: ContractEvent, *args: Any, **kwargs: Any
+    ) -> None:
+        """
+        The ``Log`` class is available to help with testing for contract events,
+        and the contents of the emitted logs.
+        """
+        self.event = contract_event()
+        try:
+            self.args = merge_args_and_kwargs(self.event.abi, args, kwargs=kwargs)
+        except TypeError:
+            raise LogError("Invalid args/kwargs provided in Log constructor.")
+        self.kwargs = kwargs
+
+    def is_present(self, receipt: TxReceipt) -> bool:
+        """
+        Asserts that *every* member of ``args`` / ``kwargs`` is present in the emitted
+        log dictionary values.
+
+        .. code:: python
+
+           assert Log(ping.events.Ping, b"one").is_present(receipt)
+           assert Log(ping.events.Ping, first=b"one").is_present(receipt)
+           assert Log(ping.events.Ping, first=b"one", second=b"two").is_present(receipt)
+           assert Log(ping.events.Ping, b"missing").is_present(receipt) is False
+           assert Log(ping.events.Ping, second=b"one").is_present(receipt) is False
+           assert Log(ping.events.Ping, b"one", b"missing").is_present(receipt) is False
+
+        """
+        logs = self._process_receipt(receipt)
+        missing_args = [arg for arg in self.args if arg not in logs.values()]
+        if missing_args:
+            return False
+        return True
+
+    def not_present(self, receipt: TxReceipt) -> bool:
+        """
+        Asserts that *every* member of ``args`` / ``kwargs`` are *not* present in the
+        emitted log dictionary values.
+
+        .. code:: python
+
+           assert Log(ping.events.Ping, b"missing").not_present(receipt)
+           assert Log(ping.events.Ping, first=b"missing").not_present(receipt)
+           assert Log(ping.events.Ping, b"one").not_present(receipt) is False
+           assert Log(ping.events.Ping, first=b"one").not_present(receipt) is False
+           assert Log(ping.events.Ping, b"one", b"missing").not_present(receipt) is False
+        """
+        logs = self._process_receipt(receipt)
+        matching_args = [arg for arg in self.args if arg in logs.values()]
+        if matching_args:
+            return False
+        return True
+
+    def exact_match(self, receipt: TxReceipt) -> bool:
+        """
+        Asserts that the provided ``kwargs`` match *exactly* the emitted log dictionary.
+        Requires ``**kwargs``, and does not accept ``*args``.
```

Comment:
Looks like this docstring is outdated?

Reply:
I kept the kwarg requirement in the `exact_match` implementation, since it made sense to me that `exact_match` requires a match in both keys and values, whereas `is_present` and `not_present` are more agnostic and will use `merge_args_and_kwargs` to accept a combination.